### PR TITLE
cat-log: support workflow files and log listing

### PR DESCRIPTION
**Pull request by oliver-sanders**
> **Built Atop:** https://github.com/cylc/cylc-uiserver/tree/update-log-work

> **Sibling:** https://github.com/cylc/cylc-flow/pull/5453

* Support subscribing to other workflow log files.
* Add interface for listing log files.
* Kill the full cat-log process tree recursively when stopping subscriptions (I seemed to hit some situation where that wasn't happening).
* Unify the `workflow` and `task` arguments in the log subscription interface into a single `id` argument.

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by oliver-sanders** [comment]
Rebasing onto 1.2.x to allow this work to be released before the Vue3 upgrade.

For safekeeping the 1.3.x branch can be found here: https://github.com/cylc/cylc-uiserver/compare/master...oliver-sanders:cylc-uiserver:update-log-work-1.3?expand=1

**Comment by oliver-sanders** [comment]
(re-opened to kick-ci into action)

File: cylc/uiserver/resolvers.py [hunk 1]
```diff
@@ -492,15 +538,88 @@ async def subscription_service(
...
+    *,
+    command='cat_log',
+    id: str,  # noqa: required to match schema arg name
+    file=None,
+    **kwargs: Any,
+) -> AsyncGenerator[Any, None]:
+    """Cat Log Resolver
+    Expands workflow provided subscription query.
+    """
+    print('% init', id)
```
Comment (hjoliver) [inline cylc/uiserver/resolvers.py:609]: (Is this for debugging?)

File: cylc/uiserver/resolvers.py [hunk 2]
```diff
@@ -492,15 +538,88 @@ async def subscription_service(
...
+    tokens = Tokens(id)
+    if kwargs.get('args', False):
+        kwargs.update(kwargs.get('args', {}))
+        kwargs.pop('args')
+    resolvers: 'Resolvers' = (
+        info.context.get('resolvers')  # type: ignore[union-attr]
+    )
+    print('% start', id)
```
Comment (hjoliver) [inline cylc/uiserver/resolvers.py:617]: (debug?)

File: cylc/uiserver/resolvers.py [hunk 3]
```diff
@@ -492,15 +538,88 @@ async def subscription_service(
...
+    async for item in resolvers.subscription_service(
+        info,
+        command,
+        [tokens],
+        file
+    ):
+        yield item
+    print('% end', id)
```
Comment (hjoliver) [inline cylc/uiserver/resolvers.py:625]: (debug?)

File: setup.cfg [hunk 4]
```diff
@@ -50,7 +50,7 @@ install_requires =
     # bleeding-edge version.
     # NB: no graphene version specified; we only make light use of it in our
     # own code, so graphene-tornado's transitive version should do.
-    cylc-flow>=8.1.2, <8.2
+    cylc-flow>=8.1.3, <8.2
```
Comment (hjoliver) [inline setup.cfg:53]: (This seems to be incompatible with the sibling cylc-flow PR, I had to change it back to get cylc-uiserver to pip install on this branch).

**Review by hjoliver — APPROVED** [review]
LGTM, tested with sibling PRs.

File: setup.cfg (hunk 4 above)
Comment (oliver-sanders) [reply setup.cfg:53]: Ah, would probably need to be `>=8.1.3.dev` until released.

File: cylc/uiserver/resolvers.py [hunk 5]
```diff
@@ -332,19 +342,21 @@ async def enqueue(stream, queue):
...
         """
-        full_workflow = workflow
-        if file and task:
-            full_workflow = Tokens(f"{workflow.workflow_id}//{task}")
-        cmd = ['cylc', 'cat-log']
+        cmd: List[str] = [
+            'cylc',
+            'cat-log',
+            '-m', 't',
+            '-o',
```
Comment (wxtim) [inline cylc/uiserver/resolvers.py:354]:
```suggestion
            '--mode', 'tail',
            '--force-remote',

**Review by wxtim — APPROVED** [review]
This looks great. I've made a stylistic suggestion, but I don't care very much whether you use it or not.

**Review by wxtim — APPROVED** [review]

**Pull request merged**